util/pyclient/bft_client.py

```diff
@@ -241,3 +247,257 @@ def get_rsi_replies(self):
+        # Load my private key and certificate
+        ssl_context.load_cert_chain(client_cert_path, client_pk_path)
+        # Server hostname to be verified must be taken from create_tls_certs.sh
+        server_hostname = self.CERT_DOMAIN_FORMAT % dest_replica.id
+        while True:
+            try:
+                # Open TCP stream and connect to server
+                tcp_stream = await trio.open_tcp_stream(str(dest_replica.ip), int(dest_replica.port))
+                # Wrap this stream with SSL stream, pass server_hostname to be verified
+                ssl_stream = trio.SSLStream(tcp_stream, ssl_context, server_hostname=server_hostname, https_compatible=False)
+                # Wait for handshake to finish (we want to be on the safe side - after this we are sure 
+                # connection is open)
+                await ssl_stream.do_handshake()
+                # Success! keep stream in dictionary and break out
+                self.ssl_streams[(dest_replica.ip, dest_replica.port)] = ssl_stream
+                return
+            except (OSError, trio.BrokenResourceError):
+                await trio.sleep(0.25)
+                pass
+            continue
+
+    async def on_stopped_replicas(self, replica_ids):
+        """ BFT network inform client when a group of replicas are stopped. We relay on caller cancel scope """
+        for id in replica_ids:
+            replica = self.replicas[id]
+            addr = (replica.ip, replica.port)
+            if addr in self.ssl_streams:
+                if addr in self.ssl_streams:
+                    stream = self.ssl_streams[addr]
+                    del self.ssl_streams[addr]
+                    await stream.aclose()
+
+    async def _send_data(self, data, dest_replica):
+        """ Try to send data to dest_replica. On exception - close the SSL stream. """
+        dest_addr = (dest_replica.ip, dest_replica.port)
+        if dest_addr not in self.ssl_streams.keys():
+            # dest_replica is not connected (crushed? isolated?) 
+            return
+        # first 4 bytes include the data header (message size), then comes the data
+        data_len = len(data)
+        out_buff = bytearray(data_len.to_bytes(self.MSG_HEADER_SIZE, "little"))
+        out_buff += bytearray(data)
+        stream = self.ssl_streams[dest_addr]
+        try:
+            await stream.send_all(out_buff)
+            return True
+        except (trio.BrokenResourceError, trio.ClosedResourceError):
+            # Failed! close the stream and return failure.
+            if dest_addr in self.ssl_streams:
+                del self.ssl_streams[dest_addr]
+                await stream.aclose()
+            return False
+
+    async def _stream_recv_some(self, out_data, dest_addr, stream, num_bytes):
+        """ Try to receive data from replica. On failure - close the SSL stream. """
+        try:
+            out_data += await stream.receive_some(num_bytes)
```

Comment:
If zero bytes are ever returned we have gotten an EOF I believe, and should return false.

Reply:
@andrewjstone 
Thanks, I would add:
I can expect only 2 values to be returned:
1) Header size self.MSG_HEADER_SIZE - 4 bytes
2) Payload size
So I think I'll make it more strict - if len(out_data) < num_bytes just return False, not just for zero.

Reply:
I don't think that's how this call works, nor how TCP streams work in general. The stream can return any number of bytes > 0 and <=4 and still be correct. If it returns exactly 0 the stream has reached an EOF indicating closure from the other side. But due to windows, delays, etc.. it's possible this call returns with say, only 2 bytes. Then the remaining 2 bytes need to be read in another call to get the complete header size. 

0 is a sentinel meaning something is broken. Returning less than max bytes (4 in this case) is normal behavior.

Reply:
You are right that the stream can return any amount of bytes up to num_bytes, including 0. I've fixed it the wrong way.
But please pay attention that there is already an outer loop in the caller which calls again.